devicetypes/smartthings/inovelli-dimmer.src/inovelli-dimmer.groovy

```diff
@@ -0,0 +1,579 @@
+		} else {
+			return state.currentPreferencesState."$preference.key".value != settings."$preference.key"
+		}
+	} else {
+		return false
+	}
+}
+
+def parse(String description) {
+	def result = null
+	if (description != "updated") {
+		def cmd = zwave.parse(description)
+		if (cmd) {
+			result = zwaveEvent(cmd)
+			log.debug("'$description' parsed to $result")
+		} else {
+			log.debug("Couldn't zwave.parse '$description'")
+		}
+	}
+	log.debug "parsed '${description}' to ${result.inspect()}"
+	result
+}
+
+def handleLEDPreferenceEvent(cmd) {
+	def hueState = [name: "hue", value: "${Math.round(zwaveValueToHuePercent(cmd.scaledConfigurationValue))}"]
+	def childDni = "${device.deviceNetworkId}:4"
+	def childDevice = childDevices.find { it.deviceNetworkId == childDni }
+	childDevice?.sendEvent(hueState)
+	childDevice?.sendEvent(name: "saturation", value: "100")
+}
+
+def createChildDevice(childDthNamespace, childDthName, childDni, childComponentLabel, childComponentName) {
+	try {
+		log.debug "Creating a child device: ${childDthNamespace}, ${childDthName}, ${childDni}, ${childComponentLabel}, ${childComponentName}"
+		addChildDevice(childDthNamespace, childDthName, childDni, device.hub.id,
+			[
+				completedSetup: true,
+				label         : childComponentLabel,
+				isComponent   : true,
+				componentName : childComponentName,
+				componentLabel: childComponentLabel
+			])
+	} catch (Exception e) {
+		log.debug "Exception: ${e}"
+	}
+}
+
+def zwaveEvent(physicalgraph.zwave.commands.basicv1.BasicReport cmd) {
+	dimmerEvents(cmd)
+}
+
+def zwaveEvent(physicalgraph.zwave.commands.switchmultilevelv3.SwitchMultilevelReport cmd) {
+	dimmerEvents(cmd)
+}
+
+def dimmerEvents(physicalgraph.zwave.Command cmd) {
+	def switchEvent = createEvent([name: "switch", value: cmd.value ? "on" : "off", descriptionText: "$device.displayName was turned ${cmd.value ? "on" : "off"}"])
+	def dimmerEvent = createEvent([name: "level", value: cmd.value == 99 ? 100 : cmd.value, unit: "%"])
+	def result = [switchEvent, dimmerEvent]
+	if (switchEvent.isStateChange) {
+		result << response(["delay 1000", zwave.meterV3.meterGet(scale: 2).format()])
+	}
+	return result
+}
+
+def on() {
+	encapSequence([
+		zwave.basicV1.basicSet(value: 0xFF),
+		zwave.basicV1.basicGet()
+	], 1000)
+}
+
+def off() {
+	encapSequence([
+		zwave.basicV1.basicSet(value: 0x00),
+		zwave.basicV1.basicGet()
+	], 1000)
+}
+
+def setLevel(level) {
+	log.debug "setLevel"
+	if (level > 99) level = 99
+	encapSequence([
+		zwave.basicV1.basicSet(value: level),
+		zwave.switchMultilevelV1.switchMultilevelGet()
+	], 1000)
+}
+
+def zwaveEvent(physicalgraph.zwave.commands.meterv3.MeterReport cmd) {
+	def map = [:]
+	if (cmd.meterType == 1 && cmd.scale == 0) {
+		map = [name: "energy", value: cmd.scaledMeterValue, unit: "kWh"]
+	} else if (cmd.meterType == 1 && cmd.scale == 2) {
+		map = [name: "power", value: Math.round(cmd.scaledMeterValue), unit: "W"]
+	}
+	createEvent(map)
+}
+
+private getButtonLabel() {
```

Comment:
I wonder if we should affix "Button" to these names, so that it reads as "Up button", "Down button", and "Configuration button"?

Reply:
Done